src/Npgsql/NpgsqlBinaryImporter.cs

```diff
@@ -301,9 +301,11 @@ public void Complete()
                 _buf.EndCopyMode();
 
                 _connector.SendMessage(CopyDoneMessage.Instance);
-                Expect<CommandCompleteMessage>(_connector.ReadMessage());
+                var cmdComplete = Expect<CommandCompleteMessage>(_connector.ReadMessage());
                 Expect<ReadyForQueryMessage>(_connector.ReadMessage());
                 _state = ImporterState.Committed;
+                // Number of rows copied on success - see "Outputs" section of https://www.postgresql.org/docs/current/static/sql-copy.html
+                return (int)cmdComplete.Rows;
```

Comment:
Couldn't this cast throw? 

@roji More broadly, is there any reason to believe that the protocol restricts this to 2^32?

https://github.com/npgsql/npgsql/blob/a0889292f45e998fab8a7aff28ae2d3318a7600c/src/Npgsql/BackendMessages/CommandCompleteMessage.cs#L32-L36


Per [the protocol docs](https://www.postgresql.org/docs/current/static/protocol-message-formats.html):

```
CommandComplete (B)
  Byte1('C')
    Identifies the message as a command-completed response.

  Int32
    Length of message contents in bytes, including self.

  String
    The command tag. This is usually a single word that identifies which SQL command was completed.

    For an INSERT command, the tag is INSERT oid rows, where rows is the number of rows inserted. oid is the object ID of the inserted row if rows is 1 and the target table has OIDs; otherwise oid is 0.

    For a DELETE command, the tag is DELETE rows where rows is the number of rows deleted.

    For an UPDATE command, the tag is UPDATE rows where rows is the number of rows updated.

    For a SELECT or CREATE TABLE AS command, the tag is SELECT rows where rows is the number of rows retrieved.

    For a MOVE command, the tag is MOVE rows where rows is the number of rows the cursor's position has been changed by.

    For a FETCH command, the tag is FETCH rows where rows is the number of rows that have been retrieved from the cursor.

    For a COPY command, the tag is COPY rows where rows is the number of rows copied. (Note: the row count appears only in PostgreSQL 8.2 and later.)
```

Reply:
Hmm, that's a good question.

There are other places in the wire protocol where the number of rows is an Int32 (e.g. the max number of rows to return in the Execute message). This doesn't prove that here it's the same but it's an indication.

Also, the ADO.NET API's [`DbDataReader.RecordsAffected`](https://docs.microsoft.com/en-us/dotnet/api/system.data.common.dbdatareader.recordsaffected?redirectedfrom=MSDN&view=netframework-4.7.2) is an `int`. This doesn't mean that PostgreSQL returns an int, but the standard .NET API does... And it seems to make sense to align the COPY API with this.

We can test, or wait until someone complains about inserting too many rows :)

Reply:
I'll see if I can find some time (and space....) to test this for the sake of curiosity. 

In API terms, I was picturing something like `LongRecordsAffected` to mirror an array's `Length` and `LongLength` members.

Reply:
OK. In any case it makes sense for me to leave this returning `int`, and if ever someone complains we can always add a `CompleteLong`.

Reply:
Wouldnt it just be easier to change the return of this method to `long` or `uint` now?

Reply:
@say25 Generally speaking, there's value in maintaining conformance with ADO.NET APIs.

That being said, since this is still open, we could just add `CompleteLong()` here, rather than punting for a future exception. But we first need to verify that the protocol _can_ return an out-of-range result.

Reply:
Continuing this [below](https://github.com/npgsql/npgsql/pull/2115#issuecomment-418475825).